packages/yoroi-extension/app/stores/ada/AdaAddressesStore.js

```diff
@@ -54,8 +58,102 @@ export async function filterMangledAddresses(request: {|
   }));
 }
 
+export type DomainResolverResponseError = 'forbidden' | 'unexpected';
+
+export type DomainResolverResponse = {|
+  nameServer: string,
+  address: ?string,
+  error: ?DomainResolverResponseError,
+|};
+
+export type DomainResolverFunc = string => Promise<?DomainResolverResponse>;
+
+export function resolveAddressDomainNameServerName(nameServerTag: string): string {
+  switch (nameServerTag) {
+    case Resolver.NameServer.Handle: return 'ADA Handle'
+    case Resolver.NameServer.Cns: return 'CNS'
+    case Resolver.NameServer.Unstoppable: return 'Unstoppable Domains'
+    default: return nameServerTag
+  }
+}
+
 export default class AdaAddressesStore extends Store<StoresMap, ActionsMap> {
 
+  _domainResolverApi: ?{| getCardanoAddresses: ({| resolve: string |}) => Promise<any> |} = null;
+
+  setup(): void {
+    super.setup();
+    this._domainResolverApi = resolverApiMaker({
+      apiConfig: {
+        [Resolver.NameServer.Unstoppable]: {
+          apiKey: 'czsajliz-wxgu6tujd1zqq7hey_pclfqhdjsqolsxjfsurgh',
+        },
+      },
+      cslFactory: RustModule.CrossCsl.init,
+    });
+  }
+
+  domainResolverSupported(): boolean {
+    const selectedWallet = this.stores.wallets.selected;
+    if (selectedWallet == null) {
+      return true;
+    }
+    return selectedWallet.getParent().getNetworkInfo().NetworkId === networks.CardanoMainnet.NetworkId;
+  }
+
+  getSupportedAddressDomainBannerState(): boolean {
+    const selectedWallet = this.stores.wallets.selected;
+    if (selectedWallet == null) {
+      return true;
+    }
+    const id = String(selectedWallet.publicDeriverId);
+    return !this.api.localStorage.getFlag(`SupportedAddressDomainBannerState/${id}/closed`);
+  }
+
+  setSupportedAddressDomainBannerState(isDisplayed: boolean): void {
+    const selectedWallet = this.stores.wallets.selected;
+    if (selectedWallet == null) {
+      return;
+    }
+    const id = String(selectedWallet.publicDeriverId);
+    this.api.localStorage.setFlag(`SupportedAddressDomainBannerState/${id}/closed`, !isDisplayed);
+  }
+
+  async resolveDomainAddress(resolve: string): Promise<?DomainResolverResponse> {
+    const { getCardanoAddresses } = this._domainResolverApi ?? {};
+    if (getCardanoAddresses == null || !isResolvableDomain(resolve)) {
+      return Promise.resolve(null);
+    }
+    const res = await getCardanoAddresses({ resolve });
+    let resultSuccess: ?DomainResolverResponse = null;
+    let resultForbidden: ?DomainResolverResponse = null;
+    let resultUnexpected: ?DomainResolverResponse = null;
```

Comment:
Why don't use just a single variable for the result?

Reply:
@nistadev , the different result types are returned in a specific order, with multiple variables for the different types the order of the return is then clearly and obviously defined in the return itself, i.e.:
```
return resultSuccess ?? resultForbidden ?? resultUnexpected;
```

This was specifically chosen for the most readability. Replacing it with a single variable would mean the priority order would need to be defined in the IF checks, like `if (result == null || result.errorType === 'unexpected')`, etc. Which means the order logic gets spread across multiple unrelated lines (separate checks in each IF) and makes it MUCH harder to understand and follow.

After your feedback I have realised that `resultSuccess` is not needed to be stored, tho, because it is always highest priority, so I have made a change to short-cut on success and return it right away. But for the two error types there are still two variables used, for the reasons of readability.